### ISSUE TITLE
Test broken by fa7fb04db4d72b9d

### ISSUE THREAD

**Issue by dreiss** [mozilla/rr#1680]
On my development server, at revision fa7fb04db4d72b9d, I got 100% reproducible failures for `thread_exit_signal` and `set_tid_address` (both with and without syscallbuf).  I could not reproduce on the previous revision, or on my laptop.  Both are 64-bit.

`set_tid_address` was failing during record with 

```
FAILED: errno=11 (Resource temporarily unavailable)
set_tid_address: /home/dreiss/local/rr/src/test/set_tid_address.c:26: main: Assertion `"FAILED: !" && check_cond(0 == syscall(202, &v, 0, 1, ((void *)0), ((void *)0), 0))' failed.
```

This suggests that the thread created on line 25 of set_tid_address.c is running to completion before the futex syscall starts.  When futex starts, v has already been set to zero, so futex fails with EAGAIN.  I ran the "rr record" command on both my laptop and server with RR_LOG=Scheduler and found this...

laptop (working):
```
[Scheduler] Scheduling next task
[Scheduler]   (29307 is un-switchable at (none))
[Scheduler]   and running; waiting for state change
[Scheduler]   new status is 0x857f
[Scheduler] Scheduling next task
[Scheduler]   29307 is blocked on SYSCALL: clone; checking status ...
[Scheduler]   ready with status 0x857f
[Scheduler]   Carrying on with task 29307
[Scheduler] Scheduling next task
[Scheduler]   (29307 is un-switchable at SYSCALL: clone)
[Scheduler] Scheduling next task
[Scheduler]   29307 isn't blocked
[Scheduler]   Carrying on with task 29307
[Scheduler] Scheduling next task
[Scheduler]   (29307 is un-switchable at (none))
[Scheduler]   and running; waiting for state change
[Scheduler]   new status is 0x7057f
[Scheduler] Scheduling next task
[Scheduler]   (29307 is un-switchable at (none))
[Scheduler]   and running; waiting for state change
[Scheduler]   new status is 0x857f
[Scheduler] Scheduling next task
[Scheduler]   (29307 is un-switchable at (none))
[Scheduler]   and running; waiting for state change
[Scheduler]   new status is 0x7057f
[Scheduler] Scheduling next task
[Scheduler]   (29307 is un-switchable at (none))
[Scheduler]   and running; waiting for state change
[Scheduler]   new status is 0x857f
[Scheduler] Scheduling next task
[Scheduler]   29307 is blocked on SYSCALL: futex; checking status ...
[Scheduler]   still blocked
[Scheduler]   need to reschedule
[Scheduler]   29308 isn't blocked
[Scheduler]   selecting task 29308
[Scheduler] Switching from 29307(set_tid_address) to 29308(set_tid_address) (priority 0 to 0) at 180
```

When it checks the status of 29307, clone is not blocked, so it's able to start the futex syscall before running the new thread.

server (not working):
```
[Scheduler] Scheduling next task
[Scheduler]   (29307 is un-switchable at (none))
[Scheduler]   and running; waiting for state change
[Scheduler]   new status is 0x857f
[Scheduler] Scheduling next task
[Scheduler]   29307 is blocked on SYSCALL: clone; checking status ...
[Scheduler]   still blocked
[Scheduler]   need to reschedule
[Scheduler]   29308 isn't blocked
[Scheduler]   selecting task 29308
[Scheduler] Switching from 29307(set_tid_address) to 29308(set_tid_address) (priority 0 to 0) at 175
```
(note that I modified the task numbers in this log to match the laptop log for easier diffing.

Any idea why clone would still be blocked on one machine, but continue right away on another?  My next step will be to try to get the contents of /proc/PID/stack when the scheduler sees clone as blocked.  I'm not sure whether that will be reliable, though.

`thread_exit_signal` was failing during replay with

```
[FATAL /home/dreiss/local/rr/src/ReplaySession.cc:423:guard_unexpected_signal() errno: 0 'Success'] 
 (task 1221071 (rec:1219426) at time 316)
 -> Assertion `false' failed to hold. Replay got unrecorded event SIGNAL: SIGSEGV(async) while awaiting signal
Launch gdb with
  gdb /tmp/rr-test-thread_exit_signal-aIdgc8K9s/thread_exit_signal-aIdgc8K9s
```

I wasn't able to get any useful information by connecting to the emergency debug server, but I did attach gdb to rr itself and got this stack:

```
#4  0x000000000061c4e6 in rr::emergency_debug (t=0x998300) at /home/dreiss/local/rr/src/log.cc:267
#5  0x000000000061c6c0 in rr::EmergencyDebugOstream::~EmergencyDebugOstream (this=0x7fff583d21f0, __in_chrg=<optimized out>) at /home/dreiss/local/rr/src/log.cc:287
#6  0x000000000067be59 in rr::guard_unexpected_signal (t=0x998300) at /home/dreiss/local/rr/src/ReplaySession.cc:423
#7  0x000000000067cd18 in rr::ReplaySession::emulate_async_signal (this=0x980fd0, t=0x998300, constraints=..., ticks=0) at /home/dreiss/local/rr/src/ReplaySession.cc:681
#8  0x000000000067e40d in rr::ReplaySession::try_one_trace_step (this=0x980fd0, t=0x998300, constraints=...) at /home/dreiss/local/rr/src/ReplaySession.cc:990
#9  0x000000000067f050 in rr::ReplaySession::replay_step (this=0x980fd0, constraints=...) at /home/dreiss/local/rr/src/ReplaySession.cc:1189
#10 0x0000000000679c44 in rr::ReplaySession::replay_step (this=0x980fd0, command=rr::RUN_CONTINUE) at /home/dreiss/local/rr/src/ReplaySession.h:259
#11 0x0000000000678c1e in rr::serve_replay_no_debugger (trace_dir="", flags=...) at /home/dreiss/local/rr/src/ReplayCommand.cc:270
#12 0x0000000000679039 in rr::replay (trace_dir="", flags=...) at /home/dreiss/local/rr/src/ReplayCommand.cc:350
```

and this information about the pending signal

```
(gdb) frame 7
#7  0x000000000067cd18 in rr::ReplaySession::emulate_async_signal (this=0x980fd0, t=0x998300, constraints=..., ticks=0) at /home/dreiss/local/rr/src/ReplaySession.cc:681
681	      guard_unexpected_signal(t);
(gdb) p t->pending_sig()
$1 = 11
(gdb) p t->pending_siginfo
$2 = {
  si_signo = 11, 
  si_errno = 0, 
  si_code = 128, 
  _sifields = {
    _pad = {0, 0, 113, 0, -503875560, 32568, -503875560, 32568, -503881280, 32568, 353, 0, -503875320, 32568, -503875320, 32568, -1, 0, 1886221359, 762475055, 1953719668, 
      1919448109, 1600414053, 1953069157, 1734964063, 762077550, 1734625633, 961230947}, 
    _kill = {
      si_pid = 0, 
      si_uid = 0
    }, 
    _timer = {
      si_tid = 0, 
      si_overrun = 0, 
      si_sigval = {
        sival_int = 113, 
        sival_ptr = 0x71
      }
    }, 
    _rt = {
      si_pid = 0, 
      si_uid = 0, 
      si_sigval = {
        sival_int = 113, 
        sival_ptr = 0x71
      }
    }, 
    _sigchld = {
      si_pid = 0, 
      si_uid = 0, 
      si_status = 113, 
      si_utime = 139882285987864, 
      si_stime = 139882285987864
    }, 
    _sigfault = {
      si_addr = 0x0
    }, 
    _sigpoll = {
      si_band = 0, 
      si_fd = 113
    }, 
    _sigsys = {
      _call_addr = 0x0, 
      _syscall = 113, 
      _arch = 0
    }
  }
}
```

The strange thing is `si_code = 128`, which doesn't appear to be a valid value.  From my brief experiments, it seems like some fault addresses (for example, 0xdeadbeefdeadbeef) generate si_code = 128 with si_addr = 0.  I'll try to collect some more information about where the SIGSEGV came from, but I'm wondering if you have any ideas of what to look for.

**Comment by rocallahan** [comment]
> Any idea why clone would still be blocked on one machine, but continue right away on another?

From rr's point of view it's nondeterministic. We do a try_wait and the kernel may or may not have scheduled the cloning task and run far enough to generate the exit-syscall ptrace trap. We need to modify `set_tid_address` so that `run_thread` is forced to finish initializing before the main thread does `FUTEX_WAIT`, probably with a pipe write just like `run_thread2` has.

**Comment by rocallahan** [comment]
> The strange thing is si_code = 128, which doesn't appear to be a valid value.

Welcome to rr, where exploring quirky kernel behaviors is a way of life :-).

It's possible this is some kind of divergence bug. You should look at the rr dump of the trace up to and including the failing event, that often gives clues. Then I'd probably step through `emulate_async_signal` and see if we're getting the tracee task into a register + ticks state that's "close to" the target state --- exact ticks match, but maybe off in one or two specific registers. Identifying how exactly the tracee triggered a SIGSEGV is also worth doing.

**Comment by dreiss** [comment]
si_code, while not documented in the man page, appears to be SI_KERNEL.  It looks like it's being triggered on an rdtsc instruction at https://sourceware.org/git/?p=glibc.git;a=blob;f=nptl/pthread_create.c;hb=refs/tags/glibc-2.17#l237 .  I'll dig into rr to see if I can figure out why it isn't handling the rdtsc properly.

**Comment by rocallahan** [comment]
My guess is that we're not supposed to reach the RDTSC at all --- that we're supposed to deliver a SIGCHLD before we reach it, according to the trace. You could confirm this by looking at the rr dump.

**Comment by dreiss** [comment]
Yep.  Here's what I've got

Error message:
```
[FATAL /home/dreiss/local/rr/src/ReplaySession.cc:423:guard_unexpected_signal() errno: 0 'Success'] 
 (task 585017 (rec:584569) at time 316)
 -> Assertion `false' failed to hold. Replay got unrecorded event SIGNAL: SIGSEGV(async) while awaiting signal
```

dump:

````
  real_time:1099214.846483 global_time:315, event:`SYSCALL: clone' (state:ENTERING_SYSCALL) tid:584466, ticks:54251
 rax:0xffffffffffffffda rbx:0x7f6df78f3700 rcx:0xffffffffffffffff rdx:0x7f6df78f39d0 rsi:0x7f6df78f2fb0 rdi:0x3d0f00 rbp:0x7f6df80d43e0 rsp:0x7ffca7f7c798 r8:0x7f6df78f3700 r9:0x7f6df78f3700 r10:0x7f6df78f39d0 r11:0x246 r12:0x0 r13:0x7f6df78f39c0 r14:0x7f6df78f3700 r15:0x22 rip:0x7f6df7bee251 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x38 fs_base:0x7f6df8724740 gs_base:0x0
}
{
  real_time:1099214.846791 global_time:316, event:`SIGNAL: SIGCHLD(async)' tid:584569, ticks:0
 rax:0x0 rbx:0x7f6df78f3700 rcx:0x7f6df7bee251 rdx:0x7f6df78f39d0 rsi:0x7f6df78f2fb0 rdi:0x3d0f00 rbp:0x7f6df80d43e0 rsp:0x7f6df78f2fb0 r8:0x7f6df78f3700 r9:0x7f6df78f3700 r10:0x7f6df78f39d0 r11:0x246 r12:0x0 r13:0x7f6df78f39c0 r14:0x7f6df78f3700 r15:0x22 rip:0x7f6df7bee251 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x38 fs_base:0x7f6df78f3700 gs_base:0x0
}
{
  real_time:1099214.846872 global_time:317, event:`SIGNAL_HANDLER: SIGCHLD(async)' tid:584569, ticks:0
 rax:0x0 rbx:0x7f6df78f3700 rcx:0x7f6df7bee251 rdx:0x7f6df78f2b40 rsi:0x7f6df78f2c70 rdi:0x11 rbp:0x7f6df80d43e0 rsp:0x7f6df78f2b38 r8:0x7f6df78f3700 r9:0x7f6df78f3700 r10:0x7f6df78f39d0 r11:0x246 r12:0x0 r13:0x7f6df78f39c0 r14:0x7f6df78f3700 r15:0x22 rip:0x400b29 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x38 fs_base:0x7f6df78f3700 gs_base:0x0
}
{
  real_time:1099214.846979 global_time:318, event:`SYSCALL: rt_sigreturn' (state:ENTERING_SYSCALL) tid:584569, ticks:0
 rax:0xffffffffffffffda rbx:0x7f6df78f3700 rcx:0xffffffffffffffff rdx:0x7f6df78f2b40 rsi:0x7f6df78f2c70 rdi:0x11 rbp:0x7f6df80d43e0 rsp:0x7f6df78f2b40 r8:0x7f6df78f3700 r9:0x7f6df78f3700 r10:0x7f6df78f39d0 r11:0x246 r12:0x0 r13:0x7f6df78f39c0 r14:0x7f6df78f3700 r15:0x22 rip:0x7f6df7b2d679 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xf fs_base:0x7f6df78f3700 gs_base:0x0
}
{
  real_time:1099214.847313 global_time:319, event:`SYSCALL: rt_sigreturn' (state:EXITING_SYSCALL) tid:584569, ticks:0
 rax:0x0 rbx:0x7f6df78f3700 rcx:0x7f6df7bee251 rdx:0x7f6df78f39d0 rsi:0x7f6df78f2fb0 rdi:0x3d0f00 rbp:0x7f6df80d43e0 rsp:0x7f6df78f2fb0 r8:0x7f6df78f3700 r9:0x7f6df78f3700 r10:0x7f6df78f39d0 r11:0x246 r12:0x0 r13:0x7f6df78f39c0 r14:0x7f6df78f3700 r15:0x22 rip:0x7f6df7bee251 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xffffffffffffffff fs_base:0x7f6df78f3700 gs_base:0x0
}
{
  real_time:1099214.847322 global_time:320, event:`EXIT_SIGHANDLER' tid:584569, ticks:0
}
{
  real_time:1099214.847387 global_time:321, event:`SEGV_RDTSC' tid:584569, ticks:3
 rax:0x746fb36c rbx:0x7f6df78f3700 rcx:0x7f6df7bee251 rdx:0xa6a6f rsi:0x7f6df78f2fb0 rdi:0x7f6df78f3700 rbp:0x0 rsp:0x7f6df78f2f20 r8:0x7f6df78f3700 r9:0x7f6df78f3700 r10:0x7f6df78f39d0 r11:0x246 r12:0x0 r13:0x7f6df78f39c0 r14:0x7f6df78f3700 r15:0x22 rip:0x7f6df7ec0d12 eflags:0x10202 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xffffffffffffffff fs_base:0x7f6df78f3700 gs_base:0x0
}
```

You can see that the child begins executing before clone returns.  The *child* immediately gets the SIGCHLD at ticks=0, then continues to the rdtsc at ticks=3.

End of the dump, where clone returns the child's tid:
```
{
  real_time:1099214.853189 global_time:334, event:`SYSCALLBUF_RESET' tid:584569, ticks:74
}
{
  real_time:1099214.853526 global_time:335, event:`EXIT' tid:584569, ticks:74
}
{
  real_time:1099214.853870 global_time:336, event:`SYSCALL: clone' (state:EXITING_SYSCALL) tid:584466, ticks:54251
 rax:0x8eb79 rbx:0x7f6df78f3700 rcx:0xffffffffffffffff rdx:0x7f6df78f39d0 rsi:0x7f6df78f2fb0 rdi:0x3d0f00 rbp:0x7f6df80d43e0 rsp:0x7ffca7f7c798 r8:0x7f6df78f3700 r9:0x7f6df78f3700 r10:0x7f6df78f39d0 r11:0x246 r12:0x0 r13:0x7f6df78f39c0 r14:0x7f6df78f3700 r15:0x22 rip:0x7f6df7bee251 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x38 fs_base:0x7f6df8724740 gs_base:0x0
}
```

Log from replay session:
```
[ReplaySession] [line 315] 584466: replaying SYSCALL: clone; state ENTERING_SYSCALL
[ReplaySession]     (syscllbufsz:0, abrtcmt:0, locked:0)
[ReplaySession] [line 316] 584569: replaying SIGNAL: SIGCHLD(async); state  (none)
[ReplaySession] advancing 0 ticks to reach 0/0x7f6df7bee251
[ReplaySession]   not same execution point: regs differ (@0x7f6df7bee251)
[ReplaySession]     (fast-forwarding over target $ip)
[ReplaySession]   not same execution point: regs differ (@0x7f6df7bee251)
[ReplaySession]   not same execution point: regs differ (@0x7f6df7bee251)
[ReplaySession]     (SIGTRAP; stepi'd target $ip)
[ReplaySession]     breaking on target $ip
[ReplaySession]   not same execution point: -3 ticks left (@0x7f6df7bee251)
```

It replays the start of clone, then tries to advance "0 ticks" and ends up at ticks=3, which is the rdtsc.

I will continue to investigate.

**Comment by dreiss** [comment]
Interesting: the ip that ReplaySession is trying to reach (0x7f6df7bee251) is equal to %rip in both the parent and child thread.  Here's the disassembly of that section:

```
78		/* End FDE now, because in the child the unwind info will be
79		   wrong.  */
80		cfi_endproc;
81		syscall
82	
83		testq	%rax,%rax
84		jl	SYSCALL_ERROR_LABEL
85		jz	L(thread_start)
86	
87		ret
(gdb) disassemble 
Dump of assembler code for function clone:
   [...]
   0x00007f6df7bee24a <+42>:	mov    $0x38,%eax
   0x00007f6df7bee24f <+47>:	syscall 
=> 0x00007f6df7bee251 <+49>:	test   %rax,%rax
   0x00007f6df7bee254 <+52>:	jl     0x7f6df7bee295 <clone+117>
   0x00007f6df7bee256 <+54>:	je     0x7f6df7bee259 <clone+57>
   0x00007f6df7bee258 <+56>:	retq   
```

It looks like during recording, SIGCHLD was delivered to the child on the first instruction after the raw clone system call.  I'll see if I can figure out whether, during replay, the child was already past that instruction when we got `[ReplaySession] advancing 0 ticks to reach 0/0x7f6df7bee251`

**Comment by rocallahan** [comment]
The key question is why, when you're at the correct IP and tick-count, the regs still differ.

**Comment by dreiss** [comment]
The answer is rcx:

```
(gdb) break src/ReplaySession.cc:1053
Breakpoint 1 at 0x67e7b0: file /home/dreiss/local/rr/src/ReplaySession.cc, line 1053.
(gdb) condition 1 trace_frame.time() == 316
(gdb) c
Continuing.

Breakpoint 1, rr::ReplaySession::setup_replay_one_trace_frame (this=0x980f70, t=0x9aedc0) at /home/dreiss/local/rr/src/ReplaySession.cc:1053
warning: Source file is more recent than executable.
1053	  LOG(debug) << "[line " << trace_frame.time() << "] " << t->rec_tid

(gdb) p/x trace_frame.recorded_regs.u.x64regs.rcx
$7 = 0x7f6df7bee251
(gdb) p/x t->regs().u.x64regs.rcx
$8 = 0xffffffffffffffff
```

It looks like the syscall instruction is allowed to clobber rcx and r11.  Maybe rr is setting it to different values in the new thread when recording and replaying.

I will continue to investigate.

**Comment by rocallahan** [comment]
See `fixup_syscall_registers` in `Task.cc`. I hope it's understandable.

So it looks like although we've set rcx to -1 in the parent when entering `clone`, in the new task the kernel resets it to the return address when returning to userspace ... which is vaguely within the semi-documented behavior of rcx in syscall entry/exit.

So I think we need to apply `fixup_syscall_registers` during recording to fix up the new task registers after the task has been created. So the fix needs to be in `prepare_clone` (assuming you're on master; I've been reworking this code recently) under `Restore modified registers in cloned task`. You should expose `fixup_syscall_registers` as a static method of `Task` and call it from `prepare_clone` to fix up the registers in `new_task`.

**Comment by dreiss** [comment]
Yes, I was just looking at `fixup_syscall_registers`.  I will try to apply it during recording.

**Comment by dreiss** [comment]
This seems to work.  What do you think?

```diff
diff --git i/src/record_syscall.cc w/src/record_syscall.cc
index 89bf614..9df29e1 100644
--- i/src/record_syscall.cc
+++ w/src/record_syscall.cc
@@ -1889,7 +1889,7 @@ static void prepare_clone(RecordTask* t, TaskSyscallState& syscall_state) {
   new_r.set_original_syscallno(
       syscall_state.syscall_entry_registers.original_syscallno());
   new_r.set_arg1(syscall_state.syscall_entry_registers.arg1());
-  new_task->set_regs(new_r);
+  new_task->emulate_syscall_entry(new_r);
 
   /* record child id here */
   if (is_clone_syscall(r.original_syscallno(), r.arch())) {
```

**Comment by rocallahan** [comment]
Yeah that looks good. It mirrors what we do in replay_syscall.

**Comment by dreiss** [comment]
Similar to set_tid_address, I'm getting a nondeterministic failure in kill_newborn.  I'm not sure why this one is so rare (while set_tid_address was 100%), but the behavior is the same: the child runs to completion before clone returns in the parent, which appears to be the intended behavior of fa7fb04.  What do you think the best solution is?

**Comment by rocallahan** [comment]
How does `kill_newborn` fail for you? `ESRCH` in the `kill` call or do the asserts on `status` fail? Whichever way it is, we should probably just modify the test to tolerate them.

**Comment by dreiss** [comment]
Oh, sorry.  The status is 0 because the child has run to completion.  Here's the output and dump

```
$ _RR_TRACE_DIR=/tmp/rr-test-kill_newborn-gPXYCJxes ./bin/rr replay
[...]
(rr) c
Continuing.
[...]...FAILED: errno=13 (Permission denied)
kill_newborn-gPXYCJxes: rr/src/test/kill_newborn.c:23: main: Assertion `"FAILED: !" && check_cond((((signed char) ((((__extension__ (((union { __typeof(status) __in; int __i; }) { .__in = (status) }).__i))) & 0x7f) + 1) >> 1) > 0))' failed.

Program received signal SIGABRT, Aborted.
0x0000000070000002 in ?? ()
(rr) bt
[...]
#12 0x0000000000400cba in main () at rr/src/test/kill_newborn.c:23
(rr) frame 12
#12 0x0000000000400cba in main () at rr/src/test/kill_newborn.c:23
23	  test_assert(WIFSIGNALED(status));
```

```
{
  real_time:1113312.588037 global_time:248, event:`SYSCALL: clone' (state:ENTERING_SYSCALL) tid:1622891, ticks:50820
 rax:0xffffffffffffffda rbx:0x38 rcx:0xffffffffffffffff rdx:0x0 rsi:0x0 rdi:0x1200011 rbp:0x7ffd3154ccb0 rsp:0x7ffd3154c9c8 r8:0x7ff0dbb8d740 r9:0x18c36b r10:0x7ff0dbb8da10 r11:0x246 r12:0x7ffd3154cc30 r13:0x7ffd3154cc50 r14:0x0 r15:0x7ffd3154cae8 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x38 fs_base:0x7ff0dbb8d740 gs_base:0x0
}
[...]
{
  real_time:1113312.597313 global_time:261, event:`SYSCALL: exit_group' (state:ENTERING_SYSCALL) tid:1622951, ticks:1275
 rax:0xffffffffffffffda rbx:0xe7 rcx:0xffffffffffffffff rdx:0x0 rsi:0x0 rdi:0x0 rbp:0x7ff0db31a840 rsp:0x7ffd3154ca50 r8:0x3c r9:0xe7 r10:0xffffffffffffff60 r11:0x246 r12:0x7ff0db31a840 r13:0x7ff0db31fe80 r14:0x0 r15:0x7ffd3154cb70 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xe7 fs_base:0x7ff0dbb8d740 gs_base:0x0
}
{
  real_time:1113312.597762 global_time:262, event:`EXIT' tid:1622951, ticks:1275
}
{
  real_time:1113312.598282 global_time:263, event:`SYSCALL: clone' (state:EXITING_SYSCALL) tid:1622891, ticks:50820
 rax:0x18c3a7 rbx:0x38 rcx:0xffffffffffffffff rdx:0x0 rsi:0x0 rdi:0x1200011 rbp:0x7ffd3154ccb0 rsp:0x7ffd3154c9c8 r8:0x7ff0dbb8d740 r9:0x18c36b r10:0x7ff0dbb8da10 r11:0x246 r12:0x7ffd3154cc30 r13:0x7ffd3154cc50 r14:0x0 r15:0x7ffd3154cae8 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x38 fs_base:0x7ff0dbb8d740 gs_base:0x0
}
{
  real_time:1113312.598437 global_time:264, event:`SYSCALL: setpriority' (state:ENTERING_SYSCALL) tid:1622891, ticks:50846
 rax:0xffffffffffffffda rbx:0x8d rcx:0xffffffffffffffff rdx:0x0 rsi:0x0 rdi:0x0 rbp:0x7ffd3154ccf0 rsp:0x7ffd3154ca50 r8:0x7ff0dbb8d740 r9:0x18c36b r10:0x7ff0dbb8da10 r11:0x246 r12:0x400940 r13:0x7ffd3154cdd0 r14:0x0 r15:0x7ffd3154cb70 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x8d fs_base:0x7ff0dbb8d740 gs_base:0x0
}
{
  real_time:1113312.599286 global_time:265, event:`SYSCALL: setpriority' (state:EXITING_SYSCALL) tid:1622891, ticks:50846
 rax:0xfffffffffffffff3 rbx:0x8d rcx:0xffffffffffffffff rdx:0x0 rsi:0x0 rdi:0x0 rbp:0x7ffd3154ccf0 rsp:0x7ffd3154ca50 r8:0x7ff0dbb8d740 r9:0x18c36b r10:0x7ff0dbb8da10 r11:0x246 r12:0x400940 r13:0x7ffd3154cdd0 r14:0x0 r15:0x7ffd3154cb70 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x8d fs_base:0x7ff0dbb8d740 gs_base:0x0
}
{
  real_time:1113312.599611 global_time:266, event:`PATCH_SYSCALL' tid:1622891, ticks:51011
}
{
  real_time:1113312.599699 global_time:267, event:`SIGNAL: SIGCHLD(async)' tid:1622891, ticks:51011
 rax:0x3e rbx:0x0 rcx:0xffffffffffffffff rdx:0x0 rsi:0x9 rdi:0x18c3a7 rbp:0x7ffd3154ccf0 rsp:0x7ffd3154ccb8 r8:0x7ff0dbb8d740 r9:0x18c36b r10:0x7ffd3154ca40 r11:0x246 r12:0x400940 r13:0x7ffd3154cdd0 r14:0x0 r15:0x0 rip:0x7ff0daf998c5 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xffffffffffffffff fs_base:0x7ff0dbb8d740 gs_base:0x0
}
{
  real_time:1113312.599711 global_time:268, event:`SIGNAL_DELIVERY: SIGCHLD(async)' tid:1622891, ticks:51011
 rax:0x3e rbx:0x0 rcx:0xffffffffffffffff rdx:0x0 rsi:0x9 rdi:0x18c3a7 rbp:0x7ffd3154ccf0 rsp:0x7ffd3154ccb8 r8:0x7ff0dbb8d740 r9:0x18c36b r10:0x7ffd3154ca40 r11:0x246 r12:0x400940 r13:0x7ffd3154cdd0 r14:0x0 r15:0x0 rip:0x7ff0daf998c5 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xffffffffffffffff fs_base:0x7ff0dbb8d740 gs_base:0x0
}
{
  real_time:1113312.599852 global_time:269, event:`SYSCALL: kill' (state:ENTERING_SYSCALL) tid:1622891, ticks:51012
 rax:0xffffffffffffffda rbx:0x3e rcx:0xffffffffffffffff rdx:0x0 rsi:0x9 rdi:0x18c3a7 rbp:0x7ffd3154ccf0 rsp:0x7ffd3154ca50 r8:0x7ff0dbb8d740 r9:0x18c36b r10:0x7ffd3154ca40 r11:0x246 r12:0x400940 r13:0x7ffd3154cdd0 r14:0x0 r15:0x7ffd3154cb70 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x3e fs_base:0x7ff0dbb8d740 gs_base:0x0
}
{
  real_time:1113312.600700 global_time:270, event:`SYSCALL: kill' (state:EXITING_SYSCALL) tid:1622891, ticks:51012
 rax:0x0 rbx:0x3e rcx:0xffffffffffffffff rdx:0x0 rsi:0x9 rdi:0x18c3a7 rbp:0x7ffd3154ccf0 rsp:0x7ffd3154ca50 r8:0x7ff0dbb8d740 r9:0x18c36b r10:0x7ffd3154ca40 r11:0x246 r12:0x400940 r13:0x7ffd3154cdd0 r14:0x0 r15:0x7ffd3154cb70 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x3e fs_base:0x7ff0dbb8d740 gs_base:0x0
}
{
  real_time:1113312.600939 global_time:271, event:`PATCH_SYSCALL' tid:1622891, ticks:51162
}
{
  real_time:1113312.601078 global_time:272, event:`SYSCALL: write' (state:ENTERING_SYSCALL) tid:1622891, ticks:51169
 rax:0xffffffffffffffda rbx:0x0 rcx:0xffffffffffffffff rdx:0x1 rsi:0x400e1f rdi:0x1 rbp:0x1 rsp:0x7ffd3154ca50 r8:0x7ff0dbb8d740 r9:0x18c36b r10:0x7ffd3154ca40 r11:0x246 r12:0x400e1f r13:0x1 r14:0x0 r15:0x7ffd3154cb70 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x1 fs_base:0x7ff0dbb8d740 gs_base:0x0
}
```

I can make the test tolerate this behavior, but then it's not really testing that rr handles killing a newborn thread.  Let me know what you'd prefer.

**Comment by rocallahan** [comment]
I think the test should tolerate that schedule. I don't see how to guarantee the schedule that we want. Since the schedule we want does happen sometimes, I think that'll have to be good enough.
